### ISSUE TITLE
Feature: A mode that run tests in all effected packages but publishes a patch only when there are code changes

### ISSUE THREAD

**Issue by osher** [lerna/lerna#1119]
Following the close of #1117.
The premise of the project is - **you publish a package only when there are changes in it.**

When a dependency is changed but it has no effect on the depending package
I expect tests to run on the depending package, but the depending package should not be published.

## Expected Behavior

- Consider a lerna project of type `independent`
- Consider a project with two packages:  `dal` and `web`, which depends on `dal`.
- Consider `dal` to be on version `1.1.6`
- Consider `web` to depend on `"dal":"^1.1.0",`
- Consider a change in `dal`, and no change in `web`

When running `lerna updated`
I should get 
 - `dal`.

When running `lerna publish`
I should be prompted for a version for `dal`.
**BUT - MOST IMPORTANT**
`web` should be updated only if the new version of `dal` does not satisfy the dependency policy of `web` and implies a change to it's `package.json`.

Examples:
When being prompted for a version:
 - If I chose `patch` - then only `dal` should be published.
 - If I chose `Minor` - then `web` should be updated as well **only if the dependency policy does not accept the new minor of `dal`**
 - If I chose `Major` - then `web` would be updated and published.

## Current Behavior

A new version of `web` is published even if nothing has changed in it.

## Possible Solution

 - Model the project into a dependency hierarchy.
 - Prompt the user for versions, sorted by hierarchy
 - Keep track of circular references - just for model's sake.
   In fact it does not matter where to start. A package would be published if either of it's dependency has publishes a version that does not satisfy it's policy, and hence requires a change in it's `package.json`.

## Steps to Reproduce (for bugs)

described above.

<details><summary>lerna.json</summary><p>

```json
<!-- Please paste your `lerna.json` here -->
```
</p></details>

<details><summary>lerna-debug.log</summary><p>

```txt
<!-- If you have a `lerna-debug.log` available, please paste it here -->
<!-- Otherwise, feel free to delete this <details> block -->
```
</p></details>

## Context

It's not OK to publish a version of a package if nothing changed in it.

IMHO, the current behavior is valid for `--exact`. Invalid without it.

In our flow publishing a version requires update the CHANGELOG.md. If there is nothing to be said there - then it's a waste of resources and needless complication of version-dependency analysis to just add there a fictive line like ` - 1.1.7 - no code changes. Same as previous version`.

I would like a version to be published only when there are effective code-changes (package.json included).
But if the current policies expressed in `package.json` are good enough - I would like them to stay as they are - because that does not justify a change.

This is a deal breaker for me.
I end up able to use just `lerna bootstrap --hoist`, and `lerna script` - which feels wierd...

## Your Environment

| Executable | Version |
| ---: | :--- |
| `lerna --version` | VERSION |
| `npm --version`  | VERSION |
| `yarn --version` | VERSION |
| `node --version` | VERSION |

| OS | Version |
| --- | --- |
| NAME | VERSION |

**Comment by evocateur** [comment]
TL;DR: Versions are cheap, developer time isn't.

I hate to be flippant, but this particular issue is not up for debate. It has been hashed and rehashed so many times already, and the actual _fact_ of past breakage when we allowed only explicit updates is incontrovertible.

Changelogs for packages with dependency updates ...maybe "Dependency updates" would be a good log? It's accurate! The package actually _did_ change, because a lerna-managed dependency changed, and therefore behavior may have changed in ways that older versions cannot deal with (semver is not a law, it's just humans trying (and often failing) to make fewer mistakes).

**Comment by osher** [comment]
well, that's a pitty.

**Comment by osher** [comment]
If you keep treating standard as recommendation, and backing up behaviors that treat them like recommendations - then recommendations is all that they'll ever be.

sad to hear this from you
